### PR TITLE
fast follow refactor, handle downthread blocked commenters

### LINKED ISSUE #19878: Notifications on a thread are still sent from a blocked user
**Describe the bug**

A notification to a reply on a thread still comes through, even when the user is blocked

**To Reproduce**

1. Block a user
2. Comment on a third-party's post
3. Get the blocked user to reply to your comment
4. Recieve a notification from a blocked user

**Expected behavior**

I should not receive a notification, as I have blocked the user.

**Screenshots**

Image

**Desktop:**

- OS, version: Windows 11 Pro
- Browser, version: Google Chrome Version 113.0.5672.93 (Official Build) (64-bit)

**Additional context**

- [Original slack thread](https://forem-team.slack.com/archives/CAA200SH5/p1690983147278609)
- Tested on Canary 1
  - [Blocked user](https://canary1.forem.lol/philiptest1_367)
  - [Comment which triggered the erroneous notification](https://canary1.forem.lol/philiptest1_367/comment/35)

### PR DESCRIPTION
**Pull request by dukegreene**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [ ] Feature
- [X] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This addresses the feedback from https://github.com/forem/forem/pull/19898, where we stopped commenters from replying on a comment where the article writer was not blocking but the author of the immediate parent comment was.

In addition to cleaning up some expectations in the spec and making the relevant controller method more readable, this change also prevents a commenter multiple comments downthread of a blocking user from leaving their comment.

## Related Tickets & Documents

- Related Issue #19878
- Closes #19878

## QA Instructions, Screenshots, Recordings
- log in as admin, grab five user emails by visiting Admin -> Member Manager -> Member show pages
- log in as user 1, search for user 2’s username, visit their profile and block them
- find an article in the feed written by user 3, and have user 1 leave a comment on it.
- log in as user 4, visit the commented-on article, leave a comment on user 1's comment
- log in as user 5, visit the article, and leave a comment on user 4's comment
- log in as (blocked) user 2 and try to leave a comment under user 5’s comment on the article
- an alert should notify the user that moderation has prevented the comment from POSTing

### UI accessibility checklist
_If your PR includes UI changes, please utilize this checklist:_
- [ ] Semantic HTML implemented?
- [ ] Keyboard operability supported?
- [ ] Checked with [axe DevTools](https://www.deque.com/axe/) and addressed `Critical` and `Serious` issues?
- [ ] Color contrast tested?

_For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] What gif best describes this PR or how it makes you feel?

guy holding out "halt" hand and gesturing at ID armband to indicate the viewer is unauthorized to proceed

### CODE REVIEW

**Review by jaw6 — APPROVED** [review]

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -314,7 +314,15 @@ def permit_commenter
...
+    blocked_by_commentable_author = UserBlock.blocking?(@comment.commentable.user_id, current_user.id)
+    blocked_by_comment_author = @comment.parent && UserBlock.blocking?(@comment.parent.user_id, current_user.id)
+
+    blocked_by_commentable_author || blocked_by_comment_author || blocker_upthread?(@comment.parent)
+  end
+
+  def blocker_upthread?(comment)
+    return false unless comment&.parent
+
+    UserBlock.blocking?(comment.parent.user_id, current_user.id) || blocker_upthread?(comment.parent)
```
Comment (lightalloy) [inline app/controllers/comments_controller.rb:326]: We can avoid recursion (harder to comprehend) and multiple queries with smth like:
```ruby
thread_authors_ids = comment.ancestors.pluck(:user_id) 
UserBlock.exists?(blocker_id: thread_authors_ids, blocked_id: blocked_id)
```
(the code needs to be checked for correctness)

File: spec/requests/comments_create_spec.rb [hunk 2]
```diff
@@ -89,21 +89,42 @@ def comment_params(**kwargs)
   end
 
   context "when user is commenting on a comment by an author who has blocked the user" do
-    let(:third_party_article) { create(:article, user_id: create(:user).id) }
-
-    it "raises a ModerationUnauthorizedError to prevent the comment from saving" do
+    before do
       create(:user_block, blocker: blocker, blocked: user, config: "default")
       # Manually manage the blocked_by_count attribute that counter_culture manages in prod
       user.update(blocked_by_count: 1)
```
Comment (lightalloy) [inline spec/requests/comments_create_spec.rb:95]: Maybe `update_column` to avoid callbacks/validations? (if applicable)

**Review by lightalloy — COMMENTED** [review]
Thanks for implementing the changes :+1: 
I have left a couple of small-ish suggestions.

File: spec/requests/comments_create_spec.rb (hunk 2 above)
Comment (dukegreene) [reply spec/requests/comments_create_spec.rb]: I think that's a good and very inexpensive improvement! Thanks!

File: app/controllers/comments_controller.rb (hunk 1 above)
Comment (dukegreene) [reply app/controllers/comments_controller.rb]: Thanks for this! I was kind of worried about how the recursion would play out both in terms of readability and performance

**Comment by dukegreene** [comment]
@lightalloy thanks for your feedback, I've updated things accordingly.

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -314,7 +314,16 @@ def permit_commenter
...
 
-    UserBlock.blocking?(@comment.commentable.user_id, current_user.id) ||
-      UserBlock.blocking?(@comment.parent&.user_id, current_user.id)
+    blocked_by_commentable_author = UserBlock.blocking?(@comment.commentable.user_id, current_user.id)
+    blocked_by_comment_author = @comment.parent && UserBlock.blocking?(@comment.parent.user_id, current_user.id)
+
+    blocked_by_commentable_author || blocked_by_comment_author || blocker_upthread?(@comment.parent)
+  end
+
+  def blocker_upthread?(comment)
```
Comment (rt4914) [inline app/controllers/comments_controller.rb:323]: Optional suggestion:
Adding a comment on this method.

```
# This method checks if any user who has commented above the given comment in
# the thread (i.e., any of the comment's ancestors) has blocked the current user.
```

**Review by rt4914 — APPROVED** [review]
LGTM. Thanks.

File: app/controllers/comments_controller.rb [hunk 4]
```diff
@@ -314,7 +314,16 @@ def permit_commenter
...
     return false if current_user.blocked_by_count.zero?
 
-    UserBlock.blocking?(@comment.commentable.user_id, current_user.id) ||
-      UserBlock.blocking?(@comment.parent&.user_id, current_user.id)
+    blocked_by_commentable_author = UserBlock.blocking?(@comment.commentable.user_id, current_user.id)
+    blocked_by_comment_author = @comment.parent && UserBlock.blocking?(@comment.parent.user_id, current_user.id)
+
+    blocked_by_commentable_author || blocked_by_comment_author || blocker_upthread?(@comment.parent)
+  end
+
```
Comment (rt4914) [inline app/controllers/comments_controller.rb:322]:
```suggestion
# This method checks if any user who has commented above the given comment in
# the thread (i.e., any of the comment's ancestors) has blocked the current user.
```

**Review by rt4914 — APPROVED** [review]

File: app/controllers/comments_controller.rb (hunk 3 above)
Comment (rt4914) [reply app/controllers/comments_controller.rb:323]: @dukegreene You can ignore this comment. I have suggested same comment before but you can directly make a commit of that, if you think that's correct else you can ignore this. Anyway its fine. Thanks.

**Comment by dukegreene** [comment]
Thank for the feedback @rt4914! I will add more comments in a fast-fast-follow-follow coming soon! ;)

**Pull request merged**